### ISSUE TITLE
`undef`/`UndefInitializer()` does not have parsable `repr`

### ISSUE THREAD

**Issue by nickrobinson251** [JuliaLang/julia#33204]
```julia
julia> VERSION
v"1.2.0"

julia> repr(undef)
"array initializer with undefined values"
```

It seems needless, to me, not to have a "parseable" representation here (https://github.com/JuliaLang/julia/pull/30757) (in the sense `x == eval(Meta.parse(repr(x)))`).

We sort of workaround this in a few places, by having `Base.undef_ref_str` 
https://github.com/JuliaLang/julia/blob/90d481e6ca1f50758841d980d0810967ccc627f6/base/show.jl#L1848

As far as I can tell, we could just change the current `show` method from
https://github.com/JuliaLang/julia/blob/90d481e6ca1f50758841d980d0810967ccc627f6/base/show.jl#L3
to only be for the `text/plain` MIME type,
e.g. 
```julia
show(io::IO, ::MIME"text/plain", ::UndefInitializer) =  print(io, "array initializer with undefined 
values")
```
Since we have 
https://github.com/JuliaLang/julia/blob/4111609c2b49372d5139c0eb9119b738bf4d73d5/base/boot.jl#L401-L402
perhaps if we wanted more compact printing we could also define 
```julia
show(io::IO, ::UndefInitializer) = print(io, "undef")
```

**Comment by nickrobinson251** [comment]
Slightly separate from the issue of a parseable `repr`: we only sometimes get the more compact `Base.undef_ref_str` printing in cases we probably want it:
e.g.
 ```julia
julia> fill(undef)
0-dimensional Array{UndefInitializer,0}:
array initializer with undefined values

julia> Array{Any, 0}(undef)
0-dimensional Array{Any,0}:
#undef
```

It seem to me that something like
```julia
julia> [undef undef undef undef]
1×4 Array{UndefInitializer,2}:
 array initializer with undefined values  …  array initializer with undefined values
```
would be much clearer as
```julia
julia> [undef undef undef undef]
1×4 Array{UndefInitializer,2}:
 undef  undef  undef  undef
```

**Comment by KristofferC** [comment]
> would be much clearer as
>
> ```julia> [undef undef undef undef]
> 1×4 Array{UndefInitializer,2}:
>  undef  undef  undef  undef```

I disagree. `undef` is only to *initalize arrays* with undefined values. It is not an undef value itelf. It is not the same as `#undef` (which if not a first class Julia object).

**Comment by nickrobinson251** [comment]
> undef is only to initalize arrays with undefined values. It is not an undef value itelf. It is not the same as #undef 

That seems fair enough, and we do get a "good" display in sensible use cases, e.g.
```julia
julia> Matrix(undef, 1, 4)
1×4 Array{Any,2}:
 #undef  #undef  #undef  #undef
```

Anyway, this was an aside :)

The main issue is the second case here:
```julia
# this seems fine
julia> UndefInitializer()  
array initializer with undefined values

# expected `undef` or `UndefInitializer()`
julia> repr(UndefInitializer())
"array initializer with undefined values" 
```

**Comment by nickrobinson251** [comment]
another slightly different, but very related issue: sometimes `repr` doesn't work at all in the presence of undefined values 🙀 
```julia
julia> x = Matrix(undef, 1, 4)
1×4 Array{Any,2}:
 #undef  #undef  #undef  #undef

julia> repr(x)
ERROR: UndefVarError: typeinfo_prefix not defined
Stacktrace:
 [1] show(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Array{Any,2}) at ./REPL[1]:4
 [2] #sprint#342(::Nothing, ::Int64, ::typeof(sprint), ::Function, ::Array{Any,2}) at ./strings/io.jl:107
 [3] #sprint at ./none:0 [inlined]
 [4] #repr#343 at ./strings/io.jl:221 [inlined]
 [5] repr(::Array{Any,2}) at ./strings/io.jl:221
 [6] top-level scope at REPL[60]:1
```

**Comment by KristofferC** [comment]
What is the use case here, i.e. what are you trying to do where this is a problem?

**Comment by nickrobinson251** [comment]
I noticed this in https://github.com/JuliaLang/julia/pull/33206/, trying to fix #31481

**Comment by nickrobinson251** [comment]
my main use-case of repr in general is to answer "how do I create a value like this?"

**Comment by KristofferC** [comment]
```
c = fill(undef)
```

seems like you tried to fill an array with undefined values and mistakenly used the undef array initializer. To me, the current printing prevented a bug.

**Comment by nickrobinson251** [comment]
certainly open to suggestions on the 0-dim `undef` case  - i.e. how to print `Array{Any, 0}(undef)` -- but I think PR  #33206 may be a better place to discuss that case :)

**Comment by stevengj** [comment]
Should be easy to fix: just change `base/show.jl` from `show(io::IO, ::UndefInitializer) = print(io, "array initializer with undefined values")` to:
```jl
show(io::IO, ::UndefInitializer) = print(io, "undef")
show(io::IO, ::MIME"text/plain", ::UndefInitializer) = print(io, "array initializer with undefined values")
```
I would be in favor of this change.

**Comment by stevengj** [comment]
I would also be in favor of printing `UndefInitializer()` instead of `undef`, since it seems more informative to make it clear that it is an instance of `UndefInitializer`.
